File: Skill scan/backend/routes/export.py
"""
Export Routes

API endpoints for exporting assessments, gaps, and profiles in PDF/CSV/ZIP formats.

Author: SkillScan Team
Date: 2026-04-15
"""

import logging
import os
from functools import wraps
from datetime import datetime

from flask import Blueprint, request, jsonify, send_file
from werkzeug.exceptions import BadRequest

from backend.utils.auth import token_required, get_current_user
from backend.utils.export_generator import ExportGenerator
from backend.utils.pdf_generator import PDFReportGenerator
from backend.utils.csv_generator import CSVExporter
from backend.models import Student, Assessment, AssessmentResponse, SkillScore, LearningPlan

logger = logging.getLogger(__name__)

export_bp = Blueprint('export', __name__, url_prefix='/export')

# Initialize exporters
export_gen = ExportGenerator()
pdf_gen = PDFReportGenerator()
csv_exp = CSVExporter()


# ==================== Helper Functions ====================

def _validate_student(student_id: int) -> tuple:
    """Validate student exists and user has access"""
    try:
        student = Student.query.get(student_id)
        if not student:
            return None, {'error': 'Student not found'}, 404
        
        current_user_id = get_current_user()
        if current_user_id != student_id:
            return None, {'error': 'Unauthorized access'}, 403
        
        return student, None, None
    except Exception as e:
        logger.error(f"Error validating student: {str(e)}")
        return None, {'error': 'Validation error'}, 500


def _get_safe_filename(filename: str) -> str:
    """Generate safe filename with timestamp"""
    timestamp = datetime.utcnow().strftime('%Y%m%d_%H%M%S')
    return f"{timestamp}_{filename}"


# ==================== Assessment Export Endpoints ====================

@export_bp.route('/assessment-pdf', methods=['POST'])
@token_required
def export_assessment_pdf():
    """
    Export single assessment as PDF
    
    POST /export/assessment-pdf
    {
        "assessment_id": 123,  (required)
        "student_id": 1       (required)
    }
    
    Returns: PDF file download
    """
    try:
        logger.info("Export assessment PDF requested")
        
        data = request.get_json()
        student_id = data.get('student_id')
        assessment_id = data.get('assessment_id')
        
        if not student_id or not assessment_id:
            return jsonify({'error': 'Missing required fields'}), 400
        
        # Validate student
        student, error_resp, status_code = _validate_student(student_id)
        if error_resp:
            return jsonify(error_resp), status_code
        
        # Verify assessment belongs to student
        assessment_response = AssessmentResponse.query.filter_by(
            id=assessment_id,
            student_id=student_id
        ).first()
        
        if not assessment_response:
            return jsonify({'error': 'Assessment not found'}), 404
        
        # Generate export data
        assessment_data = export_gen.generate_assessment_export(
            student_id,
            assessment_id
        )
        
        if not assessment_data:
            return jsonify({'error': 'Failed to generate assessment data'}), 500
        
        # Generate PDF
        pdf_bytes = pdf_gen.generate_assessment_pdf(assessment_data)
        
        if not pdf_bytes:
            return jsonify({'error': 'Failed to generate PDF'}), 500
        
        filename = _get_safe_filename('assessment_report.pdf')
        
        logger.info(f"Assessment PDF exported successfully: {filename}")
        
        return send_file(
            io.BytesIO(pdf_bytes),
            mimetype='application/pdf',
            as_attachment=True,
            download_name=filename
        )
        
    except Exception as e:
        logger.error(f"Error exporting assessment PDF: {str(e)}")
        return jsonify({'error': 'Export failed'}), 500


@export_bp.route('/assessments-csv', methods=['POST'])
@token_required
def export_assessments_csv():
    """
    Export all assessments as CSV
    
    POST /export/assessments-csv
    {
        "student_id": 1,      (required)
        "skill_id": 5         (optional)
    }
    
    Returns: CSV file download
    """
    try:
        logger.info("Export assessments CSV requested")
        
        data = request.get_json()
        student_id = data.get('student_id')
        skill_id = data.get('skill_id')
        
        if not student_id:
            return jsonify({'error': 'Missing student_id'}), 400
        
        # Validate student
        student, error_resp, status_code = _validate_student(student_id)
        if error_resp:
            return jsonify(error_resp), status_code
        
        # Get assessments
        query = AssessmentResponse.query.filter_by(student_id=student_id)
        if skill_id:
            query = query.join(Assessment).filter(Assessment.skill_id == skill_id)
        
        assessments = query.all()
        
        if not assessments:
            return jsonify({'error': 'No assessments found'}), 404
        
        # Generate export data
        assessments_data = []
        for assessment in assessments:
            data = export_gen.generate_assessment_export(student_id, assessment.id)
            if data:
                assessments_data.append(data)
        
        # Generate CSV
        csv_data = CSVExporter.export_assessments_csv(assessments_data)
        
        if not csv_data:
            return jsonify({'error': 'Failed to generate CSV'}), 500
        
        filename = _get_safe_filename('assessments.csv')
        
        logger.info(f"Assessments CSV exported: {filename}")
        
        return send_file(
            io.BytesIO(csv_data.encode('utf-8')),
            mimetype='text/csv',
            as_attachment=True,
            download_name=filename
        )
        
    except Exception as e:
        logger.error(f"Error exporting assessments CSV: {str(e)}")
        return jsonify({'error': 'Export failed'}), 500


# ==================== Gap Analysis Export Endpoints ====================

@export_bp.route('/gap-report-pdf', methods=['POST'])
@token_required
def export_gap_report_pdf():
    """
    Export gap analysis report as PDF
    
    POST /export/gap-report-pdf
    {
        "student_id": 1,      (required)
        "skill_id": 5         (required)
    }
    
    Returns: PDF file download
    """
    try:
        logger.info("Export gap report PDF requested")
        
        data = request.get_json()
        student_id = data.get('student_id')
        skill_id = data.get('skill_id')
        
        if not student_id or not skill_id:
            return jsonify({'error': 'Missing required fields'}), 400
        
        # Validate student
        student, error_resp, status_code = _validate_student(student_id)
        if error_resp:
            return jsonify(error_resp), status_code
        
        # Generate export data
        gap_data = export_gen.generate_gap_export(student_id, skill_id)
        
        if not gap_data:
            return jsonify({'error': 'Failed to generate gap data'}), 500
        
        # Generate PDF
        pdf_bytes = pdf_gen.generate_gap_report_pdf(gap_data)
        
        if not pdf_bytes:
            return jsonify({'error': 'Failed to generate PDF'}), 500
        
        filename = _get_safe_filename('gap_report.pdf')
        
        logger.info(f"Gap report PDF exported: {filename}")
        
        return send_file(
            io.BytesIO(pdf_bytes),
            mimetype='application/pdf',
            as_attachment=True,
            download_name=filename
        )
        
    except Exception as e:
        logger.error(f"Error exporting gap report PDF: {str(e)}")
        return jsonify({'error': 'Export failed'}), 500


# ==================== Profile Export Endpoints ====================

@export_bp.route('/profile-pdf', methods=['POST'])
@token_required
def export_profile_pdf():
    """
    Export complete profile as PDF
    
    POST /export/profile-pdf
    {
        "student_id": 1       (required)
    }
    
    Returns: PDF file download
    """
    try:
        logger.info("Export profile PDF requested")
        
        data = request.get_json()
        student_id = data.get('student_id')
        
        if not student_id:
            return jsonify({'error': 'Missing student_id'}), 400
        
        # Validate student
        student, error_resp, status_code = _validate_student(student_id)
        if error_resp:
            return jsonify(error_resp), status_code
        
        # Generate export data
        profile_data = export_gen.generate_profile_export(student_id)
        
        if not profile_data:
            return jsonify({'error': 'Failed to generate profile data'}), 500
        
        # Generate PDF
        pdf_bytes = pdf_gen.generate_profile_pdf(profile_data)
        
        if not pdf_bytes:
            return jsonify({'error': 'Failed to generate PDF'}), 500
        
        filename = _get_safe_filename('profile_complete.pdf')
        
        logger.info(f"Profile PDF exported: {filename}")
        
        return send_file(
            io.BytesIO(pdf_bytes),
            mimetype='application/pdf',
            as_attachment=True,
            download_name=filename
        )
        
    except Exception as e:
        logger.error(f"Error exporting profile PDF: {str(e)}")
        return jsonify({'error': 'Export failed'}), 500


@export_bp.route('/skills-csv', methods=['POST'])
@token_required
def export_skills_csv():
    """
    Export all skills as CSV
    
    POST /export/skills-csv
    {
        "student_id": 1       (required)
    }
    
    Returns: CSV file download
    """
    try:
        logger.info("Export skills CSV requested")
        
        data = request.get_json()
        student_id = data.get('student_id')
        
        if not student_id:
            return jsonify({'error': 'Missing student_id'}), 400
        
        # Validate student
        student, error_resp, status_code = _validate_student(student_id)
        if error_resp:
            return jsonify(error_resp), status_code
        
        # Generate export data
        csv_data = export_gen.generate_skills_csv(student_id)
        
        if not csv_data:
            return jsonify({'error': 'Failed to generate CSV'}), 500
        
        filename = _get_safe_filename('skills.csv')
        
        logger.info(f"Skills CSV exported: {filename}")
        
        return send_file(
            io.BytesIO(csv_data.encode('utf-8')),
            mimetype='text/csv',
            as_attachment=True,
            download_name=filename
        )
        
    except Exception as e:
        logger.error(f"Error exporting skills CSV: {str(e)}")
        return jsonify({'error': 'Export failed'}), 500


# ==================== Bulk Export Endpoints ====================

@export_bp.route('/all', methods=['POST'])
@token_required
def export_all():
    """
    Export all data as ZIP
    
    POST /export/all
    {
        "student_id": 1,       (required)
        "format": "zip"        (pdf/csv/zip)
    }
    
    Returns: ZIP file download with all exports
    """
    try:
        logger.info("Export all data requested")
        
        data = request.get_json()
        student_id = data.get('student_id')
        export_format = data.get('format', 'zip')
        
        if not student_id:
            return jsonify({'error': 'Missing student_id'}), 400
        
        # Validate student
        student, error_resp, status_code = _validate_student(student_id)
        if error_resp:
            return jsonify(error_resp), status_code
        
        # Generate profile data
        profile_data = export_gen.generate_profile_export(student_id)
        if not profile_data:
            return jsonify({'error': 'Failed to generate data'}), 500
        
        pdf_files = {}
        csv_files = {}
        
        # Generate profile PDF
        profile_pdf = pdf_gen.generate_profile_pdf(profile_data)
        if profile_pdf:
            pdf_files['profile_complete.pdf'] = profile_pdf
        
        # Generate skills CSV
        skills_csv = export_gen.generate_skills_csv(student_id)
        if skills_csv:
            csv_files['skills.csv'] = skills_csv
        
        # Generate assessments CSV
        assessments_csv = export_gen.generate_all_assessments_csv(student_id)
        if assessments_csv:
            csv_files['assessments.csv'] = assessments_csv
        
        # Create ZIP
        zip_bytes = export_gen.generate_zip_export(student_id, pdf_files, csv_files)
        
        if not zip_bytes:
            return jsonify({'error': 'Failed to create ZIP'}), 500
        
        filename = _get_safe_filename('skillscan_export_all.zip')
        
        logger.info(f"All data exported as ZIP: {filename}")
        
        return send_file(
            io.BytesIO(zip_bytes),
            mimetype='application/zip',
            as_attachment=True,
            download_name=filename
        )
        
    except Exception as e:
        logger.error(f"Error exporting all data: {str(e)}")
        return jsonify({'error': 'Export failed'}), 500


# ==================== Export Status Endpoints ====================

@export_bp.route('/status', methods=['GET'])
@token_required
def export_status():
    """
    Get export status and available options
    
    GET /export/status
    
    Returns: Available export options
    """
    try:
        logger.info("Export status requested")
        
        status = {
            'exports_available': True,
            'formats': ['pdf', 'csv', 'zip'],
            'export_types': [
                'assessment',
                'gap_report',
                'profile',
                'skills',
                'assessments',
                'all'
            ],
            'frequency_limit': 'unlimited',
            'timestamp': datetime.utcnow().isoformat()
        }
        
        return jsonify(status), 200
        
    except Exception as e:
        logger.error(f"Error getting export status: {str(e)}")
        return jsonify({'error': 'Status check failed'}), 500


# ==================== Error Handlers ====================

@export_bp.errorhandler(BadRequest)
def handle_bad_request(e):
    """Handle bad request errors"""
    logger.error(f"Bad request: {str(e)}")
    return jsonify({'error': 'Bad request'}), 400


@export_bp.errorhandler(404)
def handle_not_found(e):
    """Handle not found errors"""
    logger.error(f"Not found: {str(e)}")
    return jsonify({'error': 'Resource not found'}), 404


@export_bp.errorhandler(500)
def handle_internal_error(e):
    """Handle internal server errors"""
    logger.error(f"Internal error: {str(e)}")
    return jsonify({'error': 'Internal server error'}), 500


# ==================== Blueprint Registration ====================

def register_export_routes(app):
    """Register export routes with Flask app"""
    app.register_blueprint(export_bp)
    logger.info("Export routes registered")

File: Skill scan/frontend/src/api/export.ts
import axios, { AxiosError } from 'axios';

const API_BASE_URL = process.env.REACT_APP_API_URL || 'http://localhost:5000/api';

interface ExportResponse {
  success: boolean;
  data?: any;
  error?: string;
}

class ExportService {
  private readonly baseURL: string = `${API_BASE_URL}/export`;

  /**
   * Export single assessment as PDF
   */
  async exportAssessmentPDF(studentId: number, assessmentId: number): Promise<Blob> {
    try {
      const response = await axios.post(
        `${this.baseURL}/assessment-pdf`,
        {
          student_id: studentId,
          assessment_id: assessmentId,
        },
        {
          responseType: 'blob',
          headers: {
            'Authorization': `Bearer ${localStorage.getItem('token')}`,
          },
        }
      );

      this.triggerDownload(response.data, `assessment_${assessmentId}.pdf`);
      return response.data;
    } catch (error) {
      this.handleError(error, 'Failed to export assessment PDF');
      throw error;
    }
  }

  /**
   * Export gap report as PDF
   */
  async exportGapReportPDF(studentId: number, skillId: number): Promise<Blob> {
    try {
      const response = await axios.post(
        `${this.baseURL}/gap-report-pdf`,
        {
          student_id: studentId,
          skill_id: skillId,
        },
        {
          responseType: 'blob',
          headers: {
            'Authorization': `Bearer ${localStorage.getItem('token')}`,
          },
        }
      );

      this.triggerDownload(response.data, `gap_report_${skillId}.pdf`);
      return response.data;
    } catch (error) {
      this.handleError(error, 'Failed to export gap report');
      throw error;
    }
  }

  /**
   * Export complete profile as PDF
   */
  async exportProfilePDF(studentId: number): Promise<Blob> {
    try {
      const response = await axios.post(
        `${this.baseURL}/profile-pdf`,
        {
          student_id: studentId,
        },
        {
          responseType: 'blob',
          headers: {
            'Authorization': `Bearer ${localStorage.getItem('token')}`,
          },
        }
      );

      this.triggerDownload(response.data, 'profile_complete.pdf');
      return response.data;
    } catch (error) {
      this.handleError(error, 'Failed to export profile');
      throw error;
    }
  }

  /**
   * Export assessments as CSV
   */
  async exportAssessmentsCSV(studentId: number, skillId?: number): Promise<Blob> {
    try {
      const response = await axios.post(
        `${this.baseURL}/assessments-csv`,
        {
          student_id: studentId,
          skill_id: skillId,
        },
        {
          responseType: 'blob',
          headers: {
            'Authorization': `Bearer ${localStorage.getItem('token')}`,
          },
        }
      );

      this.triggerDownload(response.data, 'assessments.csv');
      return response.data;
    } catch (error) {
      this.handleError(error, 'Failed to export assessments CSV');
      throw error;
    }
  }

  /**
   * Export skills as CSV
   */
  async exportSkillsCSV(studentId: number): Promise<Blob> {
    try {
      const response = await axios.post(
        `${this.baseURL}/skills-csv`,
        {
          student_id: studentId,
        },
        {
          responseType: 'blob',
          headers: {
            'Authorization': `Bearer ${localStorage.getItem('token')}`,
          },
        }
      );

      this.triggerDownload(response.data, 'skills.csv');
      return response.data;
    } catch (error) {
      this.handleError(error, 'Failed to export skills CSV');
      throw error;
    }
  }

  /**
   * Export all data as ZIP
   */
  async exportAll(studentId: number, format: 'zip' | 'pdf' | 'csv' = 'zip'): Promise<Blob> {
    try {
      const response = await axios.post(
        `${this.baseURL}/all`,
        {
          student_id: studentId,
          format,
        },
        {
          responseType: 'blob',
          headers: {
            'Authorization': `Bearer ${localStorage.getItem('token')}`,
          },
        }
      );

      const filename = `skillscan_export_${new Date().toISOString().slice(0, 10)}.${
        format === 'zip' ? 'zip' : format === 'pdf' ? 'pdf' : 'csv'
      }`;

      this.triggerDownload(response.data, filename);
      return response.data;
    } catch (error) {
      this.handleError(error, 'Failed to export all data');
      throw error;
    }
  }

  /**
   * Export profile data
   */
  async exportProfile(studentId: number, format: 'pdf' | 'csv'): Promise<Blob> {
    if (format === 'pdf') {
      return this.exportProfilePDF(studentId);
    } else {
      return this.exportAssessmentsCSV(studentId);
    }
  }

  /**
   * Export assessments data
   */
  async exportAssessments(studentId: number, format: 'pdf' | 'csv'): Promise<Blob> {
    if (format === 'pdf') {
      return this.exportProfilePDF(studentId);
    } else {
      return this.exportAssessmentsCSV(studentId);
    }
  }

  /**
   * Export gap reports
   */
  async exportGapReports(studentId: number, format: 'pdf' | 'csv'): Promise<Blob> {
    // This would need a skill_id in production
    // For now, return profile which includes gaps
    return this.exportProfile(studentId, format);
  }

  /**
   * Get export status and options
   */
  async getExportStatus(): Promise<ExportResponse> {
    try {
      const response = await axios.get(
        `${this.baseURL}/status`,
        {
          headers: {
            'Authorization': `Bearer ${localStorage.getItem('token')}`,
          },
        }
      );

      return {
        success: true,
        data: response.data,
      };
    } catch (error) {
      this.handleError(error, 'Failed to get export status');
      return {
        success: false,
        error: 'Failed to get export status',
      };
    }
  }

  /**
   * Trigger file download
   */
  private triggerDownload(data: Blob, filename: string): void {
    const url = window.URL.createObjectURL(data);
    const link = document.createElement('a');
    link.href = url;
    link.setAttribute('download', filename);
    document.body.appendChild(link);
    link.click();
    link.parentNode?.removeChild(link);
    window.URL.revokeObjectURL(url);
  }

  /**
   * Handle API errors
   */
  private handleError(error: any, message: string): void {
    console.error(message, error);

    if (error instanceof AxiosError) {
      if (error.response?.status === 401) {
        throw new Error('Unauthorized. Please log in again.');
      } else if (error.response?.status === 404) {
        throw new Error('Resource not found.');
      } else if (error.response?.status === 500) {
        throw new Error('Server error. Please try again later.');
      } else if (error.message) {
        throw new Error(error.message);
      }
    }

    throw new Error(message);
  }
}

export const exportService = new ExportService();
